<a href="https://colab.research.google.com/github/cryssoga/DSWP/blob/master/Li%C3%A7%C3%A3o/Kaggle_Churn_Challenge_old.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

## Labdata Churn Challenge 2020

Try to predict whether a client will churn the telco service
Cada linha representa um cliente e cada coluna representa uma informação a respeito daquele cliente.

Os dados incluem os seguintes grupos de variáveis:

•	Churn: representa se um cliente cancelou ou não os serviços contratados

•	Serviços que cada cliente contratou - phone, multiple lines, internet, online security, online backup, device protection, tech support, and streaming TV and movies

•	Informações a respeito da conta do cliente – how long they’ve been a customer, contract, payment method, paperless billing, monthly charges, and total charges

•	Informações demográficas – gender, age range, and if they have partners and dependents

The raw data contains 7043 rows (customers) and 21 columns (features).

In [1]:
import pandas as pd
import numpy as np

In [38]:
# url dos arquivos baixados do site https://www.kaggle.com/c/labdata-churn-challenge-2020/data e uploaded em github

url_train = 'https://raw.githubusercontent.com/cryssoga/DSWP/master/Li%C3%A7%C3%A3o/train.csv'
url_test = 'https://raw.githubusercontent.com/cryssoga/DSWP/master/Li%C3%A7%C3%A3o/test.csv'
url_sample_submission = 'https://raw.githubusercontent.com/cryssoga/DSWP/master/Li%C3%A7%C3%A3o/sample_submission.csv'

In [39]:
# carregando em dataframes:

df_train = pd.read_csv(url_train)
df_test = pd.read_csv(url_test)
df_sample_submission = pd.read_csv(url_sample_submission)

In [ ]:
df_train.shape

(5634, 21)

In [ ]:
5634/21     # verificando relação entre quantidade de linhas e variáveis

268.2857142857143

In [ ]:
df_test.shape

(1409, 20)

In [ ]:
1409/(1409+5634)      # test: 20%, train: 80%

0.20005679397983814

In [ ]:
df_sample_submission.shape

(1409, 2)

In [ ]:
df_train.head(1)  

id  gender  SeniorCitizen  ... MonthlyCharges TotalCharges  Churn
2  6479  Female              0  ...          110.8       6640.7      0

[1 rows x 21 columns]

In [ ]:
df_test.head(1)

id gender  ...  MonthlyCharges TotalCharges
0  5027   Male  ...            20.0        445.3

[1 rows x 20 columns]

In [ ]:
df_sample_submission.head(1)

id  Churn
0  5027      0

In [40]:
df_train.isna().sum()

id                    0
gender                0
SeniorCitizen         0
Partner               0
Dependents          218
tenure              461
PhoneService          0
MultipleLines         0
InternetService       0
OnlineSecurity        0
OnlineBackup          0
DeviceProtection      0
TechSupport           0
StreamingTV           0
StreamingMovies       0
Contract              0
PaperlessBilling      0
PaymentMethod        99
MonthlyCharges        0
TotalCharges          0
Churn                 0
dtype: int64

In [ ]:
df_train.isna().sum().sum()

778

In [ ]:
df_train.isna().sum().sum()/df_train.shape[0]

0.13809016684416045

In [ ]:
df_train.info()


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5634 entries, 0 to 5633
Data columns (total 21 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   id                5634 non-null   int64  
 1   gender            5634 non-null   object 
 2   SeniorCitizen     5634 non-null   int64  
 3   Partner           5634 non-null   object 
 4   Dependents        5416 non-null   object 
 5   tenure            5173 non-null   float64
 6   PhoneService      5634 non-null   object 
 7   MultipleLines     5634 non-null   object 
 8   InternetService   5634 non-null   object 
 9   OnlineSecurity    5634 non-null   object 
 10  OnlineBackup      5634 non-null   object 
 11  DeviceProtection  5634 non-null   object 
 12  TechSupport       5634 non-null   object 
 13  StreamingTV       5634 non-null   object 
 14  StreamingMovies   5634 non-null   object 
 15  Contract          5634 non-null   object 
 16  PaperlessBilling  5634 non-null   object 


In [ ]:
df_train.describe()

id  SeniorCitizen       tenure  MonthlyCharges        Churn
count  5634.000000    5634.000000  5173.000000     5634.000000  5634.000000
mean   3530.313454       0.158679    32.332882       64.626438     0.265353
std    2036.195109       0.365409    24.463607       30.021806     0.441560
min       1.000000       0.000000     0.000000       18.250000     0.000000
25%    1775.250000       0.000000     9.000000       35.462500     0.000000
50%    3525.500000       0.000000    29.000000       70.250000     0.000000
75%    5303.750000       0.000000    55.000000       89.800000     1.000000
max    7043.000000       1.000000    72.000000      118.750000     1.000000

In [ ]:
df_train['Dependents'].unique()

array(['No', 'Yes', nan], dtype=object)

In [ ]:
df_train['Dependents'].value_counts()

No     3793
Yes    1623
Name: Dependents, dtype: int64

In [ ]:
df_train['Dependents'].isna().sum()/len(df_train['Dependents'])

0.038693645722399715

In [ ]:
df_train['tenure'].isna().sum()/len(df_train['tenure'])

0.08182463613773518

In [ ]:
df_train['PaymentMethod'].isna().sum()/len(df_train['PaymentMethod'])

0.01757188498402556

In [ ]:
df_test.isna().sum()

id                    0
gender                0
SeniorCitizen         0
Partner               0
Dependents           41
tenure              118
PhoneService          0
MultipleLines         0
InternetService       0
OnlineSecurity        0
OnlineBackup          0
DeviceProtection      0
TechSupport           0
StreamingTV           0
StreamingMovies       0
Contract              0
PaperlessBilling      0
PaymentMethod        30
MonthlyCharges        0
TotalCharges          0
dtype: int64

In [ ]:
df_test.isna().sum().sum()

189

In [ ]:
df_test.isna().sum().sum()/df_test.shape[0]

0.13413768630234207

In [ ]:
df_test.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1409 entries, 0 to 1408
Data columns (total 20 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   id                1409 non-null   int64  
 1   gender            1409 non-null   object 
 2   SeniorCitizen     1409 non-null   int64  
 3   Partner           1409 non-null   object 
 4   Dependents        1368 non-null   object 
 5   tenure            1291 non-null   float64
 6   PhoneService      1409 non-null   object 
 7   MultipleLines     1409 non-null   object 
 8   InternetService   1409 non-null   object 
 9   OnlineSecurity    1409 non-null   object 
 10  OnlineBackup      1409 non-null   object 
 11  DeviceProtection  1409 non-null   object 
 12  TechSupport       1409 non-null   object 
 13  StreamingTV       1409 non-null   object 
 14  StreamingMovies   1409 non-null   object 
 15  Contract          1409 non-null   object 
 16  PaperlessBilling  1409 non-null   object 


In [ ]:
df_test.describe()

id  SeniorCitizen       tenure  MonthlyCharges
count  1409.000000    1409.000000  1291.000000     1409.000000
mean   3488.757984       0.176011    32.460883       65.302520
std    2021.976356       0.380965    24.843710       30.366091
min       7.000000       0.000000     0.000000       18.700000
25%    1726.000000       0.000000     9.000000       35.750000
50%    3510.000000       0.000000    29.000000       71.000000
75%    5222.000000       0.000000    57.000000       90.100000
max    7041.000000       1.000000    72.000000      118.650000

In [ ]:
df_test['Dependents'].isna().sum()/len(df_test['Dependents'])

0.029098651525904896

In [ ]:
df_test['tenure'].isna().sum()/len(df_test['tenure'])

0.08374733853797019

In [ ]:
df_test['PaymentMethod'].isna().sum()/len(df_test['PaymentMethod'])

0.021291696238467

In [ ]:
df_train[df_train]

# Decisão: Dropar os NaNs ou não dropar?

## df_train: total de NaNs = 778

Total de linhas = 5634

Se NaNs em linhas diferentes: 5634-778 = 4856 linhas ==> perda 13,81% de dados

## df_test: total de NaNs = 189

Total de linhas = 1409

Se NaNs em linhas distintas: 1409-189 = 1220 linhas ==> perda 13,41% de dados

## Vamos dropar e verificar quantidade final de linhas:

In [ ]:
df_train_dropado = df_train.dropna()      # 4884/5634 = 86.68% ==> 13,32% perda    
df_train_dropado.shape                    # 5634 - 4884 = 750 linhas

(4884, 21)

In [ ]:
df_test_dropado = df_test.dropna()        # 1228/1409 = 87,15% ==> 12,85% perda
df_test_dropado.shape                     # 1409 - 1228 = 181 linhas

(1228, 20)

In [ ]:
df_train_dropado['Churn'].value_counts()/df_train_dropado.shape[0]

0    0.739148
1    0.260852
Name: Churn, dtype: float64

In [ ]:
df_train['Churn'].value_counts()/df_train.shape[0]

0    0.734647
1    0.265353
Name: Churn, dtype: float64

In [ ]:
df_sample_submission['Churn'].value_counts()/df_sample_submission.shape[0]

0    0.520937
1    0.479063
Name: Churn, dtype: float64

In [ ]:
df_sample_submission.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1409 entries, 0 to 1408
Data columns (total 2 columns):
 #   Column  Non-Null Count  Dtype
---  ------  --------------  -----
 0   id      1409 non-null   int64
 1   Churn   1409 non-null   int64
dtypes: int64(2)
memory usage: 22.1 KB


In [5]:
# No caso de dropar NaNs, precisa dropar linhas do df_sample_submission correspondentes:
# Criando df_test_21 => df_test (20 colunas) + coluna 'Churn'do df_sample_submission
df_test_21 = df_test.copy()
df_test_21['Churn'] = df_sample_submission['Churn']
df_test_21.head()

id  gender  SeniorCitizen  ... MonthlyCharges TotalCharges  Churn
0  5027    Male              0  ...          20.00        445.3      0
1  1733    Male              1  ...          99.00       5969.3      1
2  5384    Male              0  ...          84.75      3050.15      0
3  6554  Female              0  ...          61.45      3751.15      1
4   364  Female              0  ...          20.55        945.7      0

[5 rows x 21 columns]

In [6]:
df_test_21.shape

(1409, 21)

In [8]:
df_test_21_dropado = df_test_21.dropna()

In [10]:
df_test_21_dropado.shape

(1228, 21)

In [11]:
df_test_21_dropado['Churn'].value_counts()

0    646
1    582
Name: Churn, dtype: int64

In [12]:
df_test_21_dropado['Churn'].value_counts()/df_test_21_dropado.shape[0]

0    0.526059
1    0.473941
Name: Churn, dtype: float64

'Churn' do test com drop e sem drop:

0         0.520937     0.526059

1         0.479063     0.473941

In [ ]:
df_D = df_train[df_train['Dependents'].isna()].index

In [ ]:
df_T = df_train[df_train['tenure'].isna()].index

In [ ]:
df_P=df_train[df_train['PaymentMethod'].isna()].index

In [ ]:
df_train[df_train['PaymentMethod'].isna()]

# Decisão 2: 
Manter 'dropado' e ver no que dá?

ou

Tratar apenas variável/coluna 'tenure' que tem 461 NaNs pela mediana?

ou 

Tratar todas as 3 variáveis ('Dependents', 'tenure', e 'PaymentMethod')?

Problema em tratar 'Dependents' e 'PaymentMethod': são object, substituir pelo mais frequente?


In [16]:
df_train['PaymentMethod'].value_counts()

Electronic check             1887
Mailed check                 1272
Bank transfer (automatic)    1212
Credit card (automatic)      1164
Name: PaymentMethod, dtype: int64

In [22]:
df_train[['PaymentMethod','Churn']].value_counts().sum()

5535

In [20]:
df_train['Dependents'].value_counts()

No     3793
Yes    1623
Name: Dependents, dtype: int64

In [21]:
df_train[['Dependents','Churn']].value_counts()

Dependents  Churn
No          0        2615
Yes         0        1376
No          1        1178
Yes         1         247
dtype: int64

In [ ]:
if df_train['Dependents'].fillna[]

In [ ]:
df_train[df_train['tenure'].isna()].head(1)

id gender  SeniorCitizen  ... MonthlyCharges TotalCharges  Churn
1  6731   Male              0  ...          19.65       332.65      0

[1 rows x 21 columns]

Vendo valor de 'Churn' das linhas passíveis de drop:

---



In [ ]:
df_train[df_train['tenure'].isna()]['Churn'].value_counts()

0    327
1    134
Name: Churn, dtype: int64

In [ ]:
df_train[df_train['Dependents'].isna()]['Churn'].value_counts()

0    148
1     70
Name: Churn, dtype: int64

In [ ]:
df_train['Dependents'].unique()

In [ ]:
df_train[df_train['PaymentMethod'].isna()]['Churn'].value_counts()

0    76
1    23
Name: Churn, dtype: int64

In [ ]:
df_train['Churn'].value_counts()/df_train.shape[0]

0    0.734647
1    0.265353
Name: Churn, dtype: float64

In [27]:
df_train['tenure'].fillna(df_train['tenure'].median(), inplace=True)

In [28]:
df_train.isna().sum()

id                    0
gender                0
SeniorCitizen         0
Partner               0
Dependents          218
tenure                0
PhoneService          0
MultipleLines         0
InternetService       0
OnlineSecurity        0
OnlineBackup          0
DeviceProtection      0
TechSupport           0
StreamingTV           0
StreamingMovies       0
Contract              0
PaperlessBilling      0
PaymentMethod        98
MonthlyCharges        0
TotalCharges          0
Churn                 0
dtype: int64

In [29]:
df_train['Dependents'].fillna(df_train['Dependents'].mode(), inplace=True)

In [33]:
df_train['Dependents'].mode()

0    No
dtype: object

In [36]:
df_train['PaymentMethod'].fillna(df_train['PaymentMethod'].mode(), inplace=True)

In [37]:
df_train.isna().sum()

id                    0
gender                0
SeniorCitizen         0
Partner               0
Dependents          218
tenure                0
PhoneService          0
MultipleLines         0
InternetService       0
OnlineSecurity        0
OnlineBackup          0
DeviceProtection      0
TechSupport           0
StreamingTV           0
StreamingMovies       0
Contract              0
PaperlessBilling      0
PaymentMethod        98
MonthlyCharges        0
TotalCharges          0
Churn                 0
dtype: int64

Seguir com qual dataframe?

In [ ]:
# Dividindo df_train em df_X_train e df_y_train:
df_X_train = df_train.dropna(subset='Churn')

In [ ]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
%matplotlib inline

import statsmodels.api as sm
import statsmodels.formula.api as smf
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import roc_auc_score, roc_curve, classification_report, accuracy_score, confusion_matrix, auc